### PR TITLE
Ollama: support tools in streaming mode

### LINKED ISSUE #1971: [FEATURE] Ollama: support tools in streaming mode
Tool support has been implemented for `OllamaChatModel` in #1558.

We need to implement the same for `OllamaStreamingChatModel`.

Implementation in `OpenAiStreamingChatModel` and `OpenAiStreamingResponseBuilder` can be taken as an example.

### PR DESCRIPTION
**Pull request by bidek**
## Issue
Closes #1971

## Change
Add support for tool calling in Ollama in streaming scenario.

## General checklist
- [x] There are no breaking changes
- [x] I have added unit and integration tests for my change
- [x] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [ ] I have manually run all the unit and integration tests in the [core] (https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main] (https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green
- [ ] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [ ] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)

### CODE REVIEW

**Comment by bidek** [comment]
I don't have much experience with thread safe code, so please pay attentions to `OllamaStreamingResponseBuilder`. Is creation of simple array list sufficient ?

File: langchain4j-ollama/src/main/java/dev/langchain4j/model/ollama/OllamaStreamingResponseBuilder.java [hunk 1]
```diff
@@ -13,6 +16,7 @@ class OllamaStreamingResponseBuilder {
 
     private StringBuffer contentBuilder = new StringBuffer();
     private volatile TokenUsage tokenUsage;
+    private volatile List<ToolCall> toolsCall;
```
Comment (kpavlov) [inline langchain4j-ollama/src/main/java/dev/langchain4j/model/ollama/OllamaStreamingResponseBuilder.java]: Why not just store a reference (AtomicReference) to the last partial response object and generate Response on build() ?

File: langchain4j-ollama/src/test/java/dev/langchain4j/model/ollama/OllamaToolChatModelIT.java [hunk 2]
```diff
@@ -121,6 +137,58 @@ void should_handle_tool_without_parameter() {
...
+        // given
+        ToolExecutionResultMessage toolExecutionResultMessage = from(toolExecutionRequest, "{\"format\": \"celsius\", \"location\": \"Paris\", \"temperature\": \"32\"}");
+        List<ChatMessage> messages = asList(userMessage, aiMessage, toolExecutionResultMessage);
+
+        CompletableFuture<Response<AiMessage>> secondFutureResponse = new CompletableFuture<>();
+        streamingChatModel.generate(messages, new StreamingResponseHandler<>() {
+
+            @Override
+            public void onNext(String token) {
+            }
```
Comment (kpavlov) [inline langchain4j-ollama/src/test/java/dev/langchain4j/model/ollama/OllamaToolChatModelIT.java:170]: assertion is needed

File: langchain4j-ollama/src/test/java/dev/langchain4j/model/ollama/OllamaToolChatModelIT.java [hunk 3]
```diff
@@ -121,6 +137,58 @@ void should_handle_tool_without_parameter() {
 
     }
 
+    @Test
+    void should_handle_tools_call_in_streaming_scenario() throws Exception {
```
Comment (kpavlov) [inline langchain4j-ollama/src/test/java/dev/langchain4j/model/ollama/OllamaToolChatModelIT.java:138]: @bidek , wpuld you mind to add also a unit test, please? 
It is much easier to handle corner case scenarios with unit tests

**Review by kpavlov — COMMENTED** [review]
@bidek, thank you very much

Would you mind to create a unit test, please?

**Comment by langchain4j** [comment]
@bidek thanks a lot!

While [this PR](https://github.com/langchain4j/langchain4j/pull/1989) is still in progress, there is already a comprehensive set of integration tests that each model should pass. It would be awesome if you could checkout `new-chat-model-api` branch, open `OllamaStreamingChatModelIT` from `dev.langchain4j.model.ollama.common` package, change `supportsTools()` to return `true` and run it. Thank you in advance!🙏

File: langchain4j-ollama/src/test/java/dev/langchain4j/model/ollama/OllamaToolChatModelIT.java (hunk 3 above)
Comment (bidek) [reply langchain4j-ollama/src/test/java/dev/langchain4j/model/ollama/OllamaToolChatModelIT.java:138]: Could you elaborate on this? What kind of unit test should this be, and what should it be testing?

File: langchain4j-ollama/src/test/java/dev/langchain4j/model/ollama/OllamaToolChatModelIT.java (hunk 2 above)
Comment (bidek) [reply langchain4j-ollama/src/test/java/dev/langchain4j/model/ollama/OllamaToolChatModelIT.java:170]: I've added a counter to check if this method is invoked. Is this what you were thinking of?

File: langchain4j-ollama/src/main/java/dev/langchain4j/model/ollama/OllamaStreamingResponseBuilder.java (hunk 1 above)
Comment (bidek) [reply langchain4j-ollama/src/main/java/dev/langchain4j/model/ollama/OllamaStreamingResponseBuilder.java]: This implementation is strongly inspired by StreamingResponseBuilder and AzureOpenAiStreamingResponseBuilder.
The difference is that it seems Ollama always returns all tool calls in one response, unlike other implementations that return them in deltas.

My intention was to make it as similar as possible to other implementations.

**Comment by bidek** [comment]
@langchain4j  
I merged the current version of this branch into the specified PR.  
It fails on the following line:  

```java
assertThat(streamingMetadata2.timesOnPartialResponseWasCalled()).isGreaterThan(1);
```

For some reason, the model responds without a streaming answer and returns the entire answer instead.

Here are logs from whole comunication:
```
2024-12-07 12:59:27 [OkHttp http://localhost:32860/...] dev.langchain4j.model.ollama.OllamaRequestLoggingInterceptor.log()
DEBUG: Request:
- method: POST
- url: http://localhost:32860/api/chat
- headers: 
- body: {
  "model" : "llama3.1",
  "messages" : [ {
    "role" : "user",
    "content" : "What is the weather in Munich?"
  } ],
  "options" : {
    "temperature" : 0.0
  },
  "stream" : true,
  "tools" : [ {
    "type" : "function",
    "function" : {
      "name" : "weather",
      "parameters" : {
        "type" : "object",
        "properties" : {
          "city" : {
            "type" : "string"
          }
        },
        "required" : [ ]
      }
    }
  } ]
}
2024-12-07 12:59:29 [OkHttp http://localhost:32860/...] dev.langchain4j.model.ollama.OllamaClient.onResponse()
DEBUG: Streaming partial response: {"model":"llama3.1","created_at":"2024-12-07T11:59:29.810997963Z","message":{"role":"assistant","content":"","tool_calls":[{"function":{"name":"weather","arguments":{"city":"Munich"}}}]},"done":false}
2024-12-07 12:59:29 [OkHttp http://localhost:32860/...] dev.langchain4j.model.ollama.OllamaClient.onResponse()
DEBUG: Streaming partial response: {"model":"llama3.1","created_at":"2024-12-07T11:59:29.842415026Z","message":{"role":"assistant","content":""},"done_reason":"stop","done":true,"total_duration":2186487985,"load_duration":1581888985,"prompt_eval_count":152,"prompt_eval_duration":140000000,"eval_count":18,"eval_duration":460000000}
2024-12-07 13:01:20 [OkHttp http://localhost:32860/...] dev.langchain4j.model.ollama.OllamaRequestLoggingInterceptor.log()
DEBUG: Request:
- method: POST
- url: http://localhost:32860/api/chat
- headers: 
- body: {
  "model" : "llama3.1",
  "messages" : [ {
    "role" : "user",
    "content" : "What is the weather in Munich?"
  }, {
    "role" : "assistant",
    "tool_calls" : [ {
      "function" : {
        "name" : "weather",
        "arguments" : {
          "city" : "Munich"
        }
      }
    } ]
  }, {
    "role" : "tool",
    "content" : "sunny"
  } ],
  "options" : {
    "temperature" : 0.0
  },
  "stream" : true,
  "tools" : [ {
    "type" : "function",
    "function" : {
      "name" : "weather",
      "parameters" : {
        "type" : "object",
        "properties" : {
          "city" : {
            "type" : "string"
          }
        },
        "required" : [ ]
      }
    }
  } ]
}
2024-12-07 13:01:21 [OkHttp http://localhost:32860/...] dev.langchain4j.model.ollama.OllamaClient.onResponse()
DEBUG: Streaming partial response: {"model":"llama3.1","created_at":"2024-12-07T12:01:21.606583703Z","message":{"role":"assistant","content":"I called the weather API to get the current conditions for Munich, and it appears that it's sunny there."},"done_reason":"stop","done":true,"total_duration":644920680,"load_duration":17185139,"prompt_eval_count":93,"prompt_eval_duration":2000000,"eval_count":23,"eval_duration":624000000}

```

**Comment by langchain4j** [comment]
@bidek thanks a lot for looking into it!
Perhaps this is how Ollama handles tool calls in the streaming mode? Does it behave the same way when used via python SDK/curl?

File: langchain4j-ollama/src/main/java/dev/langchain4j/model/ollama/OllamaStreamingResponseBuilder.java (hunk 1 above)
Comment (langchain4j) [reply langchain4j-ollama/src/main/java/dev/langchain4j/model/ollama/OllamaStreamingResponseBuilder.java]: Ideally this code sohuld handle the case when Ollama will change the streaming behaviour. If could start returning token-by-token the same way as other LLM APIs do. Would be great to investigate why Ollama behaves this way, perharps there is more info on their github/discord/etc?

**Comment by dkraemerwork** [comment]
Hey @bidek for your work!
And thanks to both of you for integrating streaming tool support.

This PR is very hot right now and we really need this. We were waiting for the ollama change for while.

Is there any way I can help you guys bring this to a new release?

**Comment by langchain4j** [comment]
@dkraemerwork thank you, it would help if you could find docu and/or investigate how Ollama does streaming when calling tools. From the logs @bidek provided it seems that it behaves differently from all other LLM providers. It would be great to understand the reason for that discrepancy and if there Ollama team has plans to change that. The implementation in this PR will depend on the answer.

**Comment by bidek** [comment]
OK, I dug through Ollama issues on GitHub and found this:
https://github.com/ollama/ollama/issues/5796#issuecomment-2508764074

This confirms that this is the expected behavior.

Here is a discussion about changing this behavior sometime in the future:
https://github.com/ollama/ollama/issues/7886


Additionally, I tested it through Postman.

Sending this request with tools attached in the body.

```
{
  "model" : "llama3.1",
  "messages" : [ {
    "role" : "user",
    "content" : "What is the weather in Munich?"
  }, {
    "role" : "assistant",
    "tool_calls" : [ {
      "function" : {
        "name" : "weather",
        "arguments" : {
          "city" : "Munich"
        }
      }
    } ]
  }, {
    "role" : "tool",
    "content" : "sunny"
  } ],
  "options" : {
    "temperature" : 0.0
  },
  "stream" : true,
  "tools" : [ {
    "type" : "function",
    "function" : {
      "name" : "weather",
      "parameters" : {
        "type" : "object",
        "properties" : {
          "city" : {
            "type" : "string"
          }
        },
        "required" : [ ]
      }
    }
  } ]
}
```
Respond as:

```
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:49:29.344085844Z",
    "message": {
        "role": "assistant",
        "content": "I called the weather API to get the current conditions for Munich, and it appears that it's sunny there."
    },
    "done_reason": "stop",
    "done": true,
    "total_duration": 586547065,
    "load_duration": 13519019,
    "prompt_eval_count": 93,
    "prompt_eval_duration": 15000000,
    "eval_count": 23,
    "eval_duration": 555000000
}
```

Sending the same request without the tools section.

```
{
  "model" : "llama3.1",
  "messages" : [ {
    "role" : "user",
    "content" : "What is the weather in Munich?"
  }, {
    "role" : "assistant",
    "tool_calls" : [ {
      "function" : {
        "name" : "weather",
        "arguments" : {
          "city" : "Munich"
        }
      }
    } ]
  }, {
    "role" : "tool",
    "content" : "sunny"
  } ],
  "options" : {
    "temperature" : 0.0
  },
  "stream" : true
}
```

Response:

```
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:47.477730927Z",
    "message": {
        "role": "assistant",
        "content": "I"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:47.497021853Z",
    "message": {
        "role": "assistant",
        "content": " can"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:47.514400368Z",
    "message": {
        "role": "assistant",
        "content": "'t"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:47.533297663Z",
    "message": {
        "role": "assistant",
        "content": " provide"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:47.55636412Z",
    "message": {
        "role": "assistant",
        "content": " real"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:47.579407917Z",
    "message": {
        "role": "assistant",
        "content": "-time"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:47.60343883Z",
    "message": {
        "role": "assistant",
        "content": " weather"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:47.626355687Z",
    "message": {
        "role": "assistant",
        "content": " information"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:47.650716075Z",
    "message": {
        "role": "assistant",
        "content": "."
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:47.675544518Z",
    "message": {
        "role": "assistant",
        "content": " However"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:47.698160217Z",
    "message": {
        "role": "assistant",
        "content": ","
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:47.722693347Z",
    "message": {
        "role": "assistant",
        "content": " I"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:47.745472079Z",
    "message": {
        "role": "assistant",
        "content": " can"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:47.772706906Z",
    "message": {
        "role": "assistant",
        "content": " suggest"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:47.79514137Z",
    "message": {
        "role": "assistant",
        "content": " some"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:47.818716565Z",
    "message": {
        "role": "assistant",
        "content": " ways"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:47.841938284Z",
    "message": {
        "role": "assistant",
        "content": " for"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:47.864522395Z",
    "message": {
        "role": "assistant",
        "content": " you"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:47.889367935Z",
    "message": {
        "role": "assistant",
        "content": " to"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:47.914078558Z",
    "message": {
        "role": "assistant",
        "content": " find"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:47.938613119Z",
    "message": {
        "role": "assistant",
        "content": " out"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:47.961355539Z",
    "message": {
        "role": "assistant",
        "content": " the"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:47.985536224Z",
    "message": {
        "role": "assistant",
        "content": " current"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:48.009394229Z",
    "message": {
        "role": "assistant",
        "content": " weather"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:48.033097401Z",
    "message": {
        "role": "assistant",
        "content": " in"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:48.056922751Z",
    "message": {
        "role": "assistant",
        "content": " Munich"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:48.080678532Z",
    "message": {
        "role": "assistant",
        "content": "."
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:48.105754937Z",
    "message": {
        "role": "assistant",
        "content": " You"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:48.129069668Z",
    "message": {
        "role": "assistant",
        "content": " can"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:48.153568408Z",
    "message": {
        "role": "assistant",
        "content": " check"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:48.176941108Z",
    "message": {
        "role": "assistant",
        "content": " online"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:48.201301331Z",
    "message": {
        "role": "assistant",
        "content": " weather"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:48.225481207Z",
    "message": {
        "role": "assistant",
        "content": " websites"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:48.247903041Z",
    "message": {
        "role": "assistant",
        "content": " such"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:48.273085337Z",
    "message": {
        "role": "assistant",
        "content": " as"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:48.298424431Z",
    "message": {
        "role": "assistant",
        "content": " Acc"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:48.325796549Z",
    "message": {
        "role": "assistant",
        "content": "u"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:48.353558329Z",
    "message": {
        "role": "assistant",
        "content": "Weather"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:48.380712747Z",
    "message": {
        "role": "assistant",
        "content": " or"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:48.409736913Z",
    "message": {
        "role": "assistant",
        "content": " Weather"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:48.436618623Z",
    "message": {
        "role": "assistant",
        "content": ".com"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:48.463256747Z",
    "message": {
        "role": "assistant",
        "content": ","
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:48.493278918Z",
    "message": {
        "role": "assistant",
        "content": " or"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:48.520082193Z",
    "message": {
        "role": "assistant",
        "content": " download"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:48.545323711Z",
    "message": {
        "role": "assistant",
        "content": " a"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:48.572938945Z",
    "message": {
        "role": "assistant",
        "content": " mobile"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:48.597192291Z",
    "message": {
        "role": "assistant",
        "content": " app"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:48.62333908Z",
    "message": {
        "role": "assistant",
        "content": " like"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:48.646604033Z",
    "message": {
        "role": "assistant",
        "content": " Dark"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:48.675348344Z",
    "message": {
        "role": "assistant",
        "content": " Sky"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:48.697830726Z",
    "message": {
        "role": "assistant",
        "content": " or"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:48.723109947Z",
    "message": {
        "role": "assistant",
        "content": " Weather"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:48.746215112Z",
    "message": {
        "role": "assistant",
        "content": " Underground"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:48.772994419Z",
    "message": {
        "role": "assistant",
        "content": " that"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:48.796321595Z",
    "message": {
        "role": "assistant",
        "content": " provides"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:48.820209955Z",
    "message": {
        "role": "assistant",
        "content": " up"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:48.84337023Z",
    "message": {
        "role": "assistant",
        "content": "-to"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:48.867280957Z",
    "message": {
        "role": "assistant",
        "content": "-date"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:48.892923986Z",
    "message": {
        "role": "assistant",
        "content": " weather"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:48.916710229Z",
    "message": {
        "role": "assistant",
        "content": " forecasts"
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:48.940482408Z",
    "message": {
        "role": "assistant",
        "content": "."
    },
    "done": false
}
{
    "model": "llama3.1",
    "created_at": "2024-12-10T13:50:48.963060656Z",
    "message": {
        "role": "assistant",
        "content": ""
    },
    "done_reason": "stop",
    "done": true,
    "total_duration": 1572280055,
    "load_duration": 12088173,
    "prompt_eval_count": 46,
    "prompt_eval_duration": 43000000,
    "eval_count": 62,
    "eval_duration": 1515000000
}
```

**Comment by bidek** [comment]
To conclude:
When a tool execution request occurs, Ollama sends two "packets": one for the tool execution request and another indicating that streaming is complete.

However, when tools are attached to the request but the model does not request tool execution, Ollama sends only one "packet," even if streaming is enabled.

**Comment by langchain4j** [comment]
@bidek thanks a lot for digging into it!

> Here is a discussion about changing this behavior sometime in the future:
> https://github.com/ollama/ollama/issues/7886

Hmm, it does not seem to mention anything about returning tool calls in one vs multiple SSE events/"packets"

> However, when tools are attached to the request but the model does not request tool execution, Ollama sends only one "packet," even if streaming is enabled.

So this means that the streaming is "broken" if tools are used (even if no tool is called)?

**Comment by dkraemerwork** [comment]
Does this mean we are unable to use Streaming Responses for Tool calls using Ollama? Is there some workaround possible?

**Comment by langchain4j** [comment]
@dkraemerwork what do you mean?

This PR adds the support for tools in streaming mode

**Comment by dkraemerwork** [comment]
> @dkraemerwork what do you mean?
> 
> This PR adds the support for tools in streaming mode

It sounds like that ollama is not returning the expected streaming result-response for it to work with langchain4j. But maybe I don't see a nuance here.

**Review by ghost — APPROVED** [review]
@bidek thank you!

**Comment by langchain4j** [comment]
I think it is good enough for the first iteration, would like to include it in the next release.
But we can revisit this again
